src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/Handlers/Connection/LoginHandler.cs

```diff
@@ -0,0 +1,636 @@
+        {
+
+            PasswordChangeRequest passwordChangeRequest = context.PasswordChangeRequest;
+
+            TdsEnums.FeatureExtension requestedFeatures = TdsEnums.FeatureExtension.None;
+            FeatureExtensions features = context.Features;
+            if (context.ConnectionOptions.ConnectRetryCount > 0)
+            {
+                requestedFeatures |= TdsEnums.FeatureExtension.SessionRecovery;
+                features.SessionRecoveryRequested = true;
+            }
+
+            
+            if (ShouldRequestFedAuth(context))
+            {
+                requestedFeatures |= TdsEnums.FeatureExtension.FedAuth;
+                features.FederatedAuthenticationInfoRequested = true;
+                features.FedAuthFeatureExtensionData =
+                    new FederatedAuthenticationFeatureExtensionData
+                    {
+                        libraryType = TdsEnums.FedAuthLibrary.MSAL,
+                        authentication = context.ConnectionOptions.Authentication,
+                        fedAuthRequiredPreLoginResponse = context.FedAuthNegotiatedInPrelogin
+                    };
+            }
+
+            if (context.AccessTokenInBytes != null)
+            {
+                requestedFeatures |= TdsEnums.FeatureExtension.FedAuth;
+                features.FedAuthFeatureExtensionData = new FederatedAuthenticationFeatureExtensionData
+                {
+                    libraryType = TdsEnums.FedAuthLibrary.SecurityToken,
+                    fedAuthRequiredPreLoginResponse = context.FedAuthNegotiatedInPrelogin,
+                    accessToken = context.AccessTokenInBytes
+                };
+                // No need any further info from the server for token based authentication. So set _federatedAuthenticationRequested to true
+                features.FederatedAuthenticationRequested = true;
+            }
+
+            // The GLOBALTRANSACTIONS, DATACLASSIFICATION, TCE, and UTF8 support features are implicitly requested
+            requestedFeatures |= TdsEnums.FeatureExtension.GlobalTransactions | TdsEnums.FeatureExtension.DataClassification | TdsEnums.FeatureExtension.Tce | TdsEnums.FeatureExtension.UTF8Support;
+
+            // The SQLDNSCaching feature is implicitly set
+            requestedFeatures |= TdsEnums.FeatureExtension.SQLDNSCaching;
+
+            features.RequestedFeatures = requestedFeatures;
+
+            await TdsLogin(context, isAsync, ct).ConfigureAwait(false);
+
+            // If the workflow being used is Active Directory Authentication and server's prelogin response
+            // for FEDAUTHREQUIRED option indicates Federated Authentication is required, we have to insert FedAuth Feature Extension
+            // in Login7, indicating the intent to use Active Directory Authentication for SQL Server.
+            static bool ShouldRequestFedAuth(LoginHandlerContext context)
```

Comment:
Can you push this method also to FeatExtension types? 
e.g. FeatureExtensionFeatures.IsRequested()
Since context is already available to the FeatureExtensionFeatures, each feature can own this decision.

You can then call "_fedAuthFeature.IsRequested()" before appending the flag.